src/request/mod.rs

```diff
@@ -71,107 +68,89 @@ impl<'a, 'b> Debug for Request<'a, 'b> {
-                        return Err("No host specified in request".into())
+    pub fn from_http(req: HttpRequest, local_addr: SocketAddr, protocol: &Protocol)
+                     -> Result<Request, String> {
+        let addr = req.remote_addr().cloned();
+        let (method, uri, version, headers, body) = req.deconstruct();
+        let url = {
+            let path = uri.path();
+            let url_string = match (version, headers.get::<headers::Host>()) {
+                (_, Some(host)) => {
+                    // Attempt to prepend the Host header (mandatory in HTTP/1.1)
+                    if let Some(port) = host.port() {
+                        format!("{}://{}:{}{}", protocol.name(), host.hostname(), port, path)
+                    } else {
+                        format!("{}://{}{}", protocol.name(), host.hostname(), path)
                     }
-                };
-
-                match Url::parse(&url_string) {
-                    Ok(url) => url,
-                    Err(e) => return Err(format!("Couldn't parse requested URL: {}", e))
+                },
+                (v, None) if v < HttpVersion::Http11 => {
+                    // Attempt to use the local address? (host header is not required in HTTP/1.0).
+                    match local_addr {
+                        SocketAddr::V4(addr4) => format!("{}://{}:{}{}", protocol.name(), addr4.ip(), local_addr.port(), path),
+                        SocketAddr::V6(addr6) => format!("{}://[{}]:{}{}", protocol.name(), addr6.ip(), local_addr.port(), path),
+                    }
+                },
+                (_, None) => {
+                    return Err("No host specified in request".into())
                 }
-            },
-            _ => return Err("Unsupported request URI".into())
+            };
+
+            match Url::parse(&url_string) {
+                Ok(url) => url,
+                Err(e) => return Err(format!("Couldn't parse requested URL: {}", e))
+            }
         };
 
         Ok(Request {
             url: url,
             remote_addr: addr,
             local_addr: local_addr,
             headers: headers,
-            body: Body::new(reader),
+            body: Some(body),
             method: method,
-            extensions: TypeMap::new(),
+            extensions: TypeMap::custom(),
             version: version,
             _p: (),
         })
     }
 
+    /// Get the contents of the body as a Vec<u8>
+    ///
+    /// This consumes the body future and turns it into Vec<u8>.  Note this should not be called
+    /// from the main hyper thread, as it will potentially deadlock.
+    pub fn get_body_contents(&mut self) -> &Vec<u8> {
+        if let Some(reader) = self.body.take() {
+            let body = reader.wait().fold(Vec::new(), |mut v, input| { v.extend_from_slice(&input.unwrap()); v });
+            self.extensions.insert::<RequestBodyKey>(body);
+        }
+        self.extensions.get::<RequestBodyKey>().unwrap()
```

Comment:
I'm only just starting to get the hang of async rust, but I suspect that this will eventually want to be refactored into an async version (which stores a shared future [1] in self.extensions and returns a clone of it), and a sync version that calls the async function and waits on its result.

[1] http://alexcrichton.com/futures-rs/futures/future/trait.Future.html#method.shared